Add db setting to enable table editing

deniskaber opened the pull request:
Relates to https://linear.app/metabase/issue/WRK-96

### Description

This adds DB setting to enable table editing on db level. Feature token check is also still enabled.

### How to verify

1. Go to Admin settings > Databases > `<db name>` > "Editable tables" toggle should be visible.
2. Try to enable it
3. Check that api request with `database-enable-table-editing` db setting is sent
4. Based on this value Table editing should be enabled or disabled.
5. This should affect URL where we go from database page on a table click.

 
### Demo

https://www.loom.com/share/911d5aa1876d4891b28a73eb14d4ad25

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

deniskaber committed:
Add db setting to enable table editing

(cherry picked from commit 59d1d29d63ac27a68084897a8c0809aeae497c02)
(cherry picked from commit dbf455decccda46163979d0c232ddf75b7f031f5)

deniskaber committed:
Adapt to new design

(cherry picked from commit 3cac55139c87424cf4084d3b2432919f8b7c3acd)

deniskaber committed:
Update text

stasgavrylov commented on frontend/src/metabase/admin/databases/components/DatabaseTableEditingSection.tsx:
```diff
@@ -0,0 +1,76 @@
+      description={t`Allow the data within tables of this database to be edited by Admin users.`}
+      data-testid="database-table-editing-section"
+    >
+      <div>
```
_minor_: it seems that this `div` is unnecessary.

wotbrew committed:
Add data editing configuration checks with a flag (#55104)

* Support intended configuration checks

e.g
- database-enable-editing is required
- database-enable-actions is not required
- superuser is required

* Fix long line in docstring

* Factor out init helper

* Less lossy asserts

deniskaber committed:
Merge branch 'internal-tools' into WRK-96-toggle-table-editing-per-database

iethree commented on frontend/src/metabase/databases/utils/settings.ts:
```diff
@@ -0,0 +1,8 @@
+import type { Database } from "metabase-types/api";
+
+export function hasDatabaseTableEditingEnabled(_database: Database) {
+  return true;
+
+  // TODO [WRK]: enable this check after it is supported on the BE
+  // return database.settings?.["database-enable-table-editing"] ?? false;
```
Is there a ticket to make sure we don't forget about this?

iethree commented on frontend/src/metabase-lib/v1/metadata/Database.ts:
```diff
@@ -121,6 +121,10 @@ class Database {
     return Boolean(this.settings?.["database-enable-actions"]);
   }
 
+  hasTableEditingEnabled() {
+    return Boolean(this.settings?.["database-enable-table-editing"]);
+  }
```
why do you need this in addition to the function in [frontend/src/metabase/databases/utils/settings.ts](https://github.com/metabase/metabase/pull/54990/files#diff-c8a6c1c8312c1868db8ce58ffbc92a35d7cb898f4a7b928c3f0ec9f6de20178c) ?

sloansparger commented on frontend/src/metabase/admin/databases/components/DatabaseTableEditingSection.tsx:
```diff
@@ -0,0 +1,76 @@
+import { useState } from "react";
```
This feature is new to me, so sorry if this is a dumb question. Should this code live in the enterprise folder? I'm wondering if this file and `frontend/src/metabase/databases/utils/settings.ts` should both be values on `PLUGIN_DATA_EDITING`?

sloansparger commented on frontend/src/metabase/databases/utils/settings.ts:
```diff
@@ -0,0 +1,8 @@
+import type { Database } from "metabase-types/api";
+
+export function hasDatabaseTableEditingEnabled(_database: Database) {
```
Should this be on `PLUGIN_DATA_EDITING`? That way if they don't have the feature enabled we can default the implementation to false?

I'm wondering about the case where a user has configured a database while paying for the enterprise edition but is later downgraded to OSS. Maybe the BE already takes care of this, but we'll want to make sure that the `true` isn't respected. (Also +1 to Ryan's point in `frontend/src/metabase-lib/v1/metadata/Database.ts`, would be nice to collocate the logic).

sloansparger commented on frontend/src/metabase/admin/databases/components/DatabaseTableEditingSection.tsx:
```diff
@@ -0,0 +1,76 @@
+
+  return (
+    <DatabaseInfoSection
+      name={t`Editable tables`}
+      description={t`Allow the data within tables of this database to be edited by Admin users.`}
```
!nit the section title matching the label looks a bit weird. Maybe we can have the writing team weigh in?

deniskaber committed:
use real setting value

deniskaber committed:
Merge branch 'internal-tools' into WRK-96-toggle-table-editing-per-database

deniskaber committed:
Review fixes

deniskaber committed:
Add e2e test

deniskaber committed:
Merge branch 'internal-tools' into WRK-96-toggle-table-editing-per-database

deniskaber committed:
Fix unit naming
